### PR TITLE
Add `Conv2D` and `Linear` based decoders

### LINKED ISSUE #167: Add linear and conv2d decoder for segmentation tasks
Lets add a linear and conv2d decoder heads for 2D segmentation tasks.

This will be along with a segmentation config for `TotalSegmentor2D` task.

### PR DESCRIPTION
**Pull request by ioangatop**
Closes #167

### CODE REVIEW

File: src/eva/vision/models/networks/decoders/conv.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        features of shape (batch_size, hidden_size, num_patches_height,
+        num_patches_width), where num_patches is image_size / patch_size.
+        Note the num_patches is also known as grid_size.
+
+        Args:
+            layers: The convolutional layers to be used as the decoder head.
+        """
+        super().__init__()
+
+        self._layers = layers
```
Comment (nkaenzig) [inline src/eva/vision/models/networks/decoders/conv.py:28]: Should we apply some sanity checks here to make sure that the `layers` passed in are compatible with this decoder?

**Review by nkaenzig — COMMENTED** [review]
Looks good! I left a couple of minor comments.

File: src/eva/vision/models/networks/decoders/decoder.py
[the remark this answers is not in the document]
Comment (ioangatop) [reply src/eva/vision/models/networks/decoders/decoder.py:1]: no head is for logits, like classification task. When you recreate an "image" like its decoder (like a unet network)

[the remark this answers is not in the document]
Comment (nkaenzig) [reply src/eva/vision/models/networks/decoders/decoder.py:1]: I'd say head is a more general term, and a decoder can be a head, but not all heads are decoders?
(Not saying we shouldn't call it decoder here, I just remember that back then we had this confusion with head v.s. decoder and the naming decision back then was not obvious)

File: src/eva/vision/models/networks/decoders/conv.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+
+class ConvDecoder(decoder.Decoder):
+    """Convolutional segmentation decoder."""
+
+    def __init__(self, layers: nn.Module) -> None:
+        """Initializes the convolutional based decoder head.
+
+        Here the input nn layers will be directly applied to the
+        features of shape (batch_size, hidden_size, num_patches_height,
```
[the remark this answers is not in the document]
Comment (ioangatop) [reply src/eva/vision/models/networks/decoders/conv.py:19]: ah good suggestion - updated

File: src/eva/vision/models/networks/decoders/decoder.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+"""Semantic segmentation decoder base class."""
+
+from torch import nn
+
+
+class Decoder(nn.Module):
```
[the remark this answers is not in the document]
Comment (ioangatop) [reply src/eva/vision/models/networks/decoders/decoder.py:6]: I think its ok, bth i dont think we will have any other decoder (and the base class is super generic so it shouuld bwe fine if we add anything more)

File: src/eva/vision/models/networks/decoders/decoder.py
[the remark this answers is not in the document]
Comment (ioangatop) [reply src/eva/vision/models/networks/decoders/decoder.py:1]: afaik head is something that maps embeddings to a lower dimensionality (classification, object detection etc) but decoder is when is the same as the input (segmentation, image generation etc). I would suggest to use this convention because I think is most common and this is what we use in the other ML teams

File: src/eva/vision/models/networks/decoders/linear.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        Args:
+            features: List of multi-level image features of shape (batch_size,
+                hidden_size, num_patches_height, num_patches_width).
+
+        Returns:
+            A tensor of shape (batch_size, hidden_size, num_patches_height,
+            num_patches_width) which is feature map of the decoder head.
+        """
+        return features[-1]
```
[the remark this answers is not in the document]
Comment (ioangatop) [reply src/eva/vision/models/networks/decoders/linear.py:41]: done!

File: src/eva/vision/models/networks/decoders/conv.py (hunk 1 above)
Comment (ioangatop) [reply src/eva/vision/models/networks/decoders/conv.py:28]: I'm not sure if we can cover all the edge cases here 🤔  I think like this is ok since its documented
what do you think?
